src/AutoRest.CSharp/Common/Output/Models/Types/ObjectType.cs

```diff
@@ -3,107 +3,72 @@
 {
-    internal class ObjectType : TypeProvider
+    internal abstract class ObjectType : TypeProvider
     {
-        private readonly SerializationBuilder _serializationBuilder;
-        private readonly TypeFactory _typeFactory;
-        private readonly SchemaTypeUsage _usage;
-
-        protected ObjectTypeProperty[]? _properties;
-        private ObjectTypeDiscriminator? _discriminator;
-        protected CSharpType? _inheritsType;
-        private CSharpType? _implementsDictionaryType;
-        private ObjectSerialization[]? _serializations;
-        private readonly ModelTypeMapping? _sourceTypeMapping;
         private ObjectTypeConstructor[]? _constructors;
-        private ObjectTypeProperty? _additionalPropertiesProperty;
+        protected ObjectTypeProperty[]? _properties;
+        private CSharpType? _inheritsType;
         private ObjectTypeConstructor? _serializationConstructor;
+        private CSharpType? _implementsDictionaryType;
         private ObjectTypeConstructor? _initializationConstructor;
+        private ObjectSerialization[]? _serializations;
+        private ObjectTypeDiscriminator? _discriminator;
 
-        public ObjectType(ObjectSchema objectSchema, BuildContext context) : base(context)
+        public ObjectType(BuildContext context) : base(context)
         {
-            OjectSchema = objectSchema;
-            _typeFactory = context.TypeFactory;
-            _serializationBuilder = new SerializationBuilder();
-            _usage = context.SchemaUsageProvider.GetUsage(OjectSchema);
-
-            var hasUsage = _usage.HasFlag(SchemaTypeUsage.Model);
-
-            DefaultAccessibility = objectSchema.Extensions?.Accessibility ?? (hasUsage ? "public" : "internal");
-            Description = BuilderHelpers.CreateDescription(objectSchema);
-
-            DefaultNamespace = GetDefaultNamespace(objectSchema, context);
-            _sourceTypeMapping = context.SourceInputModel?.CreateForModel(ExistingType);
-
-            // Update usage from code attribute
-            if (_sourceTypeMapping?.Usage != null)
-            {
-                foreach (var usage in _sourceTypeMapping.Usage)
-                {
-                    _usage |= Enum.Parse<SchemaTypeUsage>(usage, true);
-                }
-            }
         }
 
-        protected ObjectSchema OjectSchema { get; }
-
-        public bool IsStruct => ExistingType?.IsValueType == true;
-        protected override string DefaultName => OjectSchema.CSharpName();
-        protected override string DefaultAccessibility { get; } = "public";
-        protected override string DefaultNamespace { get; }
-        protected override TypeKind TypeKind => IsStruct ? TypeKind.Struct : TypeKind.Class;
-
-        public string? Description { get; }
+        public ObjectTypeConstructor[] Constructors => _constructors ??= BuildConstructors().ToArray();
+        public virtual ObjectTypeProperty[] Properties => _properties ??= BuildProperties().ToArray();
```

Comment:
Both this and BuildProperties are virtual. We should pick one.

Reply:
this one is tricky due to the infinite loop.  Its easier if I can explain by showing you tomorrow.